mmaction/datasets/tube_dataset.py

```diff
@@ -0,0 +1,183 @@
+        num_classes (int): Number class of the dataset. Default: 24.
+        data_prefix (str | None): Path to a directory where videos are held.
+            Default: None.
+        test_mode (bool): Store True when building test or validation dataset.
+            Default: False.
+        encoding (str): Encode style to load pickle file.
+            Default: 'iso-8859-1'.
+        filename_tmpl (str): Template for each filename. Default: '{:05}.jpg'.
+        split (int): Index of split to indicate the data part for train and
+            test videos. Default: 1.
+        tube_length (int): Length of tube to form the tubelet. Default: 7.
+        start_index (int): Specify a start index for frames in consideration of
+            different filename format. However, when taking videos as input,
+            it should be set to 0, since frames loaded from videos count
+            from 0. Default: 1.
+        modality (str): Modality of data. Support 'RGB', 'Flow'.
+            Default: 'RGB'.
+    """
+
+    def __init__(self,
+                 ann_file,
+                 pipeline,
+                 preload_ann_file=None,
+                 save_preload=False,
+                 num_classes=24,
+                 data_prefix=None,
+                 test_mode=False,
+                 encoding='iso-8859-1',
+                 filename_tmpl='{:05}.jpg',
+                 split=1,
+                 tube_length=7,
+                 start_index=1,
+                 modality='RGB'):
+        self.preload_ann_file = preload_ann_file
+        self.save_preload = save_preload
+        self.filename_tmpl = filename_tmpl
+        self.split = split
+        self.tube_length = tube_length
+        self.encoding = encoding
+        super().__init__(
+            ann_file,
+            pipeline,
+            data_prefix,
+            test_mode=test_mode,
+            num_classes=num_classes,
+            start_index=start_index,
+            modality=modality)
+
+    @staticmethod
+    def tubelet_in_tube(video_tube, frame_index, tube_length):
+        return all([
+            i in video_tube[:, 0]
+            for i in range(frame_index, frame_index + tube_length)
+        ])
+
+    @staticmethod
+    def tubelet_out_tube(video_tube, frame_index, tube_length):
+        return all([
+            i not in video_tube[:, 0]
+            for i in range(frame_index, frame_index + tube_length)
+        ])
+
+    def check_tubelet(self, video_tubes, frame_index, tube_length):
+        is_whole = all([
+            self.tubelet_in_tube(video_tube, frame_index, tube_length)
+            or self.tubelet_out_tube(video_tube, frame_index, tube_length)
+            for video_tube in video_tubes
+        ])
+        has_gt = any([
+            self.tubelet_in_tube(tube, frame_index, tube_length)
+            for tube in video_tubes
+        ])
+        return is_whole and has_gt
+
+    def load_annotations(self):
+        pkl_data = pickle.load(
```

Comment:
Just curious, why specifying the encoding here?

Reply:
Yes, if `pickle.load` can not load the downloaded annotation file, better to convert the encoding during preprocessing.